File: trip-server/eval/multi-turn-smoke.ts
#!/usr/bin/env ts-node
/**
 * 多轮 chat cache hit rate 演化测试
 *
 * 流程：
 * 1. 登录拿 token
 * 2. 发 turn 1（无 conversationId），记录 hitRate
 * 3. 发 turn 2（带 conversationId），记录 hitRate
 * 4. 发 turn 3（带 conversationId），记录 hitRate
 *
 * 用法：
 *   EVAL_BASE_URL=http://127.0.0.1:3000 npx ts-node eval/multi-turn-smoke.ts
 *
 * 输出：每轮的 prompt / cached / hitRate，3 个城市各跑一次
 */

const BASE = process.env.EVAL_BASE_URL || 'http://127.0.0.1:3000'
const USER = process.env.EVAL_USERNAME
const PASS = process.env.EVAL_PASSWORD

interface SSEEvent { type: string; content?: string; error?: string; data?: any; usage?: any }

interface TurnResult {
  index: number
  message: string
  prompt: number
  cached: number
  hitRate: number
  durationMs: number
  error?: string
}

async function login(): Promise<string> {
  const res = await fetch(`${BASE}/api/user/login`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ username: USER, password: PASS }),
  })
  if (!res.ok) throw new Error(`login failed: ${res.status} ${await res.text()}`)
  const data = (await res.json()) as { data?: { token?: string } }
  if (!data.data?.token) throw new Error(`no token in response`)
  return data.data.token
}

async function chatOnce(token: string, message: string, conversationId?: number): Promise<{ usage?: any; conversationId?: number; error?: string; text: string }> {
  const res = await fetch(`${BASE}/api/trip/chat`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify({ message, conversationId }),
  })
  if (!res.ok || !res.body) {
    return { error: `chat failed: ${res.status}`, text: '' }
  }

  const reader = res.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''
  let text = ''
  let usage: any
  let returnedConvId: number | undefined
  let error: string | undefined

  while (true) {
    const { value, done } = await reader.read()
    if (done) break
    buffer += decoder.decode(value, { stream: true })
    let idx
    while ((idx = buffer.indexOf('\n\n')) !== -1) {
      const raw = buffer.slice(0, idx)
      buffer = buffer.slice(idx + 2)
      const dataLines: string[] = []
      for (const line of raw.split('\n')) {
        if (line.startsWith('data:')) dataLines.push(line.slice(5).trim())
      }
      if (dataLines.length === 0) continue
      const dataStr = dataLines.join('\n')
      try {
        const ev = JSON.parse(dataStr) as SSEEvent
        if (ev.type === 'chunk' && ev.content) text += ev.content
        if (ev.type === 'complete') {
          if (ev.data?.conversationId) returnedConvId = ev.data.conversationId
          if ((ev as any).usage) usage = (ev as any).usage
          else if (ev.data?.usage) usage = ev.data.usage
        }
        if (ev.type === 'error') error = ev.error || 'unknown'
      } catch {}
    }
  }
  return { usage, conversationId: returnedConvId, error, text }
}

interface Scenario {
  name: string
  turns: string[]
}

const SCENARIOS: Scenario[] = [
  {
    name: 'chengdu-3days',
    turns: [
      '帮我规划成都3日行程，带父母，慢节奏，喜欢美食和茶馆',
      '第二天能加个火锅吗，父母不太能吃辣',
      '那预算大概多少？三个人',
    ],
  },
  {
    name: 'tokyo-5days',
    turns: [
      '帮我规划东京5日行程，学生党，预算 8000 块',
      '推荐几个免费景点',
      '从机场到新宿怎么走最便宜',
    ],
  },
  {
    name: 'xian-2days-kid',
    turns: [
      '帮我规划西安2日行程，带 6 岁小孩，想看兵马俑但是怕他累',
      '第二天安排轻松点，最好有午休时间',
      '兵马俑附近有适合小孩吃的地方吗',
    ],
  },
]

async function runScenario(token: string, s: Scenario): Promise<TurnResult[]> {
  const results: TurnResult[] = []
  let conversationId: number | undefined

  for (let i = 0; i < s.turns.length; i++) {
    const message = s.turns[i]
    const start = Date.now()
    const r = await chatOnce(token, message, conversationId)
    const durationMs = Date.now() - start

    if (r.error) {
      results.push({ index: i + 1, message, prompt: 0, cached: 0, hitRate: 0, durationMs, error: r.error })
      break
    }

    if (r.conversationId) conversationId = r.conversationId

    const prompt = r.usage?.prompt ?? 0
    const cached = r.usage?.cached ?? 0
    const hitRate = prompt > 0 ? cached / prompt : 0
    results.push({ index: i + 1, message, prompt, cached, hitRate, durationMs })
  }

  return results
}

async function main() {
  console.log(`Multi-turn smoke test → ${BASE}\n`)
  const token = await login()
  console.log(`logged in.\n`)

  const allResults: Record<string, TurnResult[]> = {}
  for (const s of SCENARIOS) {
    console.log(`━━━ ${s.name} ━━━`)
    const results = await runScenario(token, s)
    allResults[s.name] = results
    for (const r of results) {
      const hitPct = (r.hitRate * 100).toFixed(1)
      const color = r.hitRate >= 0.5 ? '🟢' : r.hitRate >= 0.3 ? '🟡' : '🔴'
      if (r.error) {
        console.log(`  turn ${r.index}  ❌ ${r.error}  (${r.durationMs}ms)`)
      } else {
        console.log(`  turn ${r.index}  ${color} hitRate=${hitPct}%  prompt=${r.prompt.toLocaleString()}  cached=${r.cached.toLocaleString()}  (${r.durationMs}ms)`)
      }
    }
    console.log()
  }

  // 汇总
  console.log(`━━━ summary ━━━`)
  console.log(`scenario        | turn1    | turn2    | turn3    | avg`)
  for (const [name, rs] of Object.entries(allResults)) {
    const r1 = rs[0] ? `${(rs[0].hitRate * 100).toFixed(1)}%` : 'n/a'
    const r2 = rs[1] ? `${(rs[1].hitRate * 100).toFixed(1)}%` : 'n/a'
    const r3 = rs[2] ? `${(rs[2].hitRate * 100).toFixed(1)}%` : 'n/a'
    const valid = rs.filter(r => !r.error)
    const avg = valid.length > 0 ? (valid.reduce((s, r) => s + r.hitRate, 0) / valid.length * 100).toFixed(1) + '%' : 'n/a'
    console.log(`${name.padEnd(15)} | ${r1.padEnd(8)} | ${r2.padEnd(8)} | ${r3.padEnd(8)} | ${avg}`)
  }
}

main().catch((e) => {
  console.error(e)
  process.exit(1)
})

File: trip-server-py/eval/multi_turn_smoke.py
"""多轮 chat cache hit rate 演化测试

对标 Node.js eval/multi-turn-smoke.ts。

流程：
1. 登录拿 token
2. 发 turn 1（无 conversationId），记录 hitRate
3. 发 turn 2（带 conversationId），记录 hitRate
4. 发 turn 3（带 conversationId），记录 hitRate

用法：
  EVAL_BASE_URL=http://127.0.0.1:3000 python -m eval.multi_turn_smoke

输出：每轮的 prompt / cached / hitRate，3 个城市各跑一次
"""

from __future__ import annotations

import asyncio
import json
import os
import sys
import time
from dataclasses import dataclass
from typing import Any, Optional

import httpx

# ---------------------------------------------------------------------------
# 环境变量
# ---------------------------------------------------------------------------

BASE = os.environ.get("EVAL_BASE_URL", "http://127.0.0.1:3000")
USER = os.environ.get("EVAL_USERNAME", "")
PASS = os.environ.get("EVAL_PASSWORD", "")


# ---------------------------------------------------------------------------
# 类型
# ---------------------------------------------------------------------------


@dataclass
class SSEEvent:
    type: str = ""
    content: str | None = None
    error: str | None = None
    data: dict | None = None
    usage: dict | None = None


@dataclass
class TurnResult:
    index: int
    message: str
    prompt: int = 0
    cached: int = 0
    hit_rate: float = 0.0
    duration_ms: int = 0
    error: str | None = None


@dataclass
class Scenario:
    name: str
    turns: list[str]


# ---------------------------------------------------------------------------
# 场景定义
# ---------------------------------------------------------------------------

SCENARIOS: list[Scenario] = [
    Scenario(
        name="chengdu-3days",
        turns=[
            "帮我规划成都3日行程，带父母，慢节奏，喜欢美食和茶馆",
            "第二天能加个火锅吗，父母不太能吃辣",
            "那预算大概多少？三个人",
        ],
    ),
    Scenario(
        name="tokyo-5days",
        turns=[
            "帮我规划东京5日行程，学生党，预算 8000 块",
            "推荐几个免费景点",
            "从机场到新宿怎么走最便宜",
        ],
    ),
    Scenario(
        name="xian-2days-kid",
        turns=[
            "帮我规划西安2日行程，带 6 岁小孩，想看兵马俑但是怕他累",
            "第二天安排轻松点，最好有午休时间",
            "兵马俑附近有适合小孩吃的地方吗",
        ],
    ),
]


# ---------------------------------------------------------------------------
# SSE 解析
# ---------------------------------------------------------------------------


def _parse_sse_event(raw: str) -> SSEEvent | None:
    """解析一条 SSE 事件（多行 data: 拼接）。"""
    data_lines: list[str] = []
    for line in raw.split("\n"):
        if line.startswith("data:"):
            data_lines.append(line[5:].strip())
    if not data_lines:
        return None
    data_str = "\n".join(data_lines)
    try:
        parsed = json.loads(data_str)
        if isinstance(parsed, dict):
            return SSEEvent(
                type=parsed.get("type", ""),
                content=parsed.get("content"),
                data=parsed.get("data"),
                error=parsed.get("error"),
                usage=parsed.get("usage"),
            )
    except json.JSONDecodeError:
        return SSEEvent(type="chunk", content=data_str)
    return None


async def _parse_sse_stream(response: httpx.Response) -> dict[str, Any]:
    """解析 SSE 流，返回 usage/conversationId/error/text。"""
    text = ""
    usage: dict | None = None
    returned_conv_id: int | None = None
    error: str | None = None
    buffer = ""

    async for raw_line in response.aiter_lines():
        buffer += raw_line + "\n"
        while "\n\n" in buffer:
            idx = buffer.index("\n\n")
            raw_event = buffer[:idx]
            buffer = buffer[idx + 2 :]
            if not raw_event.strip():
                continue
            event = _parse_sse_event(raw_event)
            if event is None:
                continue
            if event.type == "chunk" and event.content:
                text += event.content
            elif event.type == "complete":
                if event.data and isinstance(event.data, dict):
                    conv_id = event.data.get("conversationId")
                    if conv_id is not None:
                        returned_conv_id = int(conv_id)
                u = event.usage or (event.data.get("usage") if event.data else None)
                if u and isinstance(u, dict):
                    usage = u
            elif event.type == "error":
                error = event.error or "unknown"

    return {"text": text, "usage": usage, "conversation_id": returned_conv_id, "error": error}


# ---------------------------------------------------------------------------
# 登录 + 单次 chat
# ---------------------------------------------------------------------------


async def login(client: httpx.AsyncClient) -> str:
    """登录拿 JWT token。"""
    resp = await client.post(
        f"{BASE}/api/user/login",
        json={"username": USER, "password": PASS},
        headers={"Content-Type": "application/json"},
    )
    if resp.status_code != 200:
        raise RuntimeError(f"login failed: {resp.status_code} {resp.text}")
    data = resp.json()
    token = (data.get("data") or {}).get("token")
    if not token:
        raise RuntimeError(f"no token in response: {data}")
    return token


async def chat_once(
    client: httpx.AsyncClient,
    token: str,
    message: str,
    conversation_id: int | None = None,
) -> dict[str, Any]:
    """调一次 chat 接口（SSE 流）。"""
    body: dict[str, Any] = {"message": message}
    if conversation_id is not None:
        body["conversationId"] = conversation_id

    async with client.stream(
        "POST",
        f"{BASE}/api/trip/chat",
        json=body,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {token}",
        },
    ) as resp:
        if resp.status_code != 200:
            await resp.aread()
            return {"error": f"chat failed: {resp.status_code}", "text": ""}
        return await _parse_sse_stream(resp)


# ---------------------------------------------------------------------------
# 跑一个场景
# ---------------------------------------------------------------------------


async def run_scenario(
    client: httpx.AsyncClient, token: str, scenario: Scenario
) -> list[TurnResult]:
    results: list[TurnResult] = []
    conversation_id: int | None = None

    for i, message in enumerate(scenario.turns):
        start = time.monotonic()
        r = await chat_once(client, token, message, conversation_id)
        duration_ms = int((time.monotonic() - start) * 1000)

        if r.get("error"):
            results.append(
                TurnResult(
                    index=i + 1,
                    message=message,
                    duration_ms=duration_ms,
                    error=r["error"],
                )
            )
            break

        if r.get("conversation_id"):
            conversation_id = r["conversation_id"]

        usage = r.get("usage") or {}
        prompt = usage.get("prompt", 0) or usage.get("input_tokens", 0)
        cached = usage.get("cached", 0) or usage.get("cache_read", 0)
        hit_rate = cached / prompt if prompt > 0 else 0.0

        results.append(
            TurnResult(
                index=i + 1,
                message=message,
                prompt=prompt,
                cached=cached,
                hit_rate=hit_rate,
                duration_ms=duration_ms,
            )
        )

    return results


# ---------------------------------------------------------------------------
# 格式化输出
# ---------------------------------------------------------------------------


def _hit_emoji(rate: float) -> str:
    if rate >= 0.5:
        return "🟢"
    if rate >= 0.3:
        return "🟡"
    return "🔴"


def print_results(all_results: dict[str, list[TurnResult]]) -> None:
    for name, results in all_results.items():
        print(f"━━━ {name} ━━━")
        for r in results:
            if r.error:
                print(f"  turn {r.index}  ❌ {r.error}  ({r.duration_ms}ms)")
            else:
                pct = f"{r.hit_rate * 100:.1f}"
                color = _hit_emoji(r.hit_rate)
                print(
                    f"  turn {r.index}  {color} hitRate={pct}%  "
                    f"prompt={r.prompt:,}  cached={r.cached:,}  "
                    f"({r.duration_ms}ms)"
                )
        print()

    # 汇总表
    print("━━━ summary ━━━")
    print("scenario        | turn1    | turn2    | turn3    | avg")
    for name, rs in all_results.items():
        cells = []
        for i in range(3):
            if i < len(rs):
                cells.append(f"{rs[i].hit_rate * 100:.1f}%")
            else:
                cells.append("n/a")
        valid = [r for r in rs if not r.error]
        avg = (
            f"{sum(r.hit_rate for r in valid) / len(valid) * 100:.1f}%"
            if valid
            else "n/a"
        )
        row = f"{name:<15} | {cells[0]:<8} | {cells[1]:<8} | {cells[2]:<8} | {avg}"
        print(row)


# ---------------------------------------------------------------------------
# main
# ---------------------------------------------------------------------------


async def amain() -> None:
    print(f"Multi-turn smoke test → {BASE}\n")
    async with httpx.AsyncClient(timeout=httpx.Timeout(120.0, connect=30.0)) as client:
        token = await login(client)
        print("logged in.\n")

        all_results: dict[str, list[TurnResult]] = {}
        for scenario in SCENARIOS:
            results = await run_scenario(client, token, scenario)
            all_results[scenario.name] = results

    print_results(all_results)


def main() -> None:
    try:
        asyncio.run(amain())
    except KeyboardInterrupt:
        print("\ninterrupted.")
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
